feat: DB-less mode (#826)
Kong/developer.konghq.com: app/gateway/db-less-mode.md

We call this DB-less mode. When running {{site.base_gateway}} DB-less, the configuration of 
entities is done in a second configuration file, in YAML or JSON, using declarative configuration.

The combination of DB-less mode and declarative configuration has a number
of benefits:

* Reduced number of dependencies: No need to manage a database installation
  if the entire setup for your use-cases fits in memory.
* Automation in CI/CD scenarios: Configuration for
  entities can be kept in a single source of truth managed via a Git
  repository.
* Enables more deployment options for {{site.base_gateway}}.

{:.warning}
> **Important**: [decK](/deck/) also manages configuration declaratively, but it requires
a database to perform any of its sync, dump, or similar operations. Therefore, decK 
can't be used in DB-less mode.

<!--vale off -->
{% mermaid %}
flowchart TD

A(<img src="/assets/logos/kogo-white.svg" style="max-height:20px" class="no-image-expand"/> {{site.base_gateway}} instance)
B(<img src="/assets/logos/kogo-white.svg" style="max-height:20px" class="no-image-expand"/> {{site.base_gateway}} instance)

> C(<img src="/assets/logos/kogo-white.svg" style="max-height:20px" class="no-image-expand"/> {{site.base_gateway}} instance)

Reviewer: I don't know if you can do anything about this, but the logo is barely visible in light mode

Author: I'll change it to the blue one

A2(fa:fa-file kong1.yml)
B2(fa:fa-file kong1.yml)
C2(fa:fa-file kong1.yml)

A2 --> A
B2 --> B
C2 --> C

style A stroke:none,fill:#0E44A2,color:#fff
style B stroke:none,fill:#0E44A2,color:#fff
style C stroke:none,fill:#0E44A2,color:#fff

{% endmermaid %}
<!-- vale on-->

## How declarative configuration works in DB-less mode

The key idea in declarative configuration is the notion
that it is *declarative*, as opposed to an *imperative* style of
configuration. "Imperative" means that a configuration is given as a series of
orders. "Declarative" means that the configuration is
given all at once.

The [Admin API](/api/gateway/admin-ee/) is an example of an imperative configuration tool. The
final state of the configuration is attained through a sequence of API calls:
one call to create a Service, another call to create a Route, another call to
add a Plugin, and so on.

> Performing the configuration incrementally like this has the undesirable

Reviewer: ```suggestion
Configuring incrementally like this has the undesirable
```

`database=off` or via environment variables. For more information, see [Set up {{site.base_gateway}} in DB-less mode](/how-to/set-up-gateway-in-dbless/).

## Load the declarative configuration file

There are two ways to load a declarative configuration file into {{site.base_gateway}}: using
`kong.conf` or the using the [`/config` Admin API endpoint](/api/gateway/admin-ee/#/operations/post-config).

You can use the following `kong.conf` parameters to load the declarative config file:

<!--vale off-->
{% kong_config_table %}
config:
  - name: declarative_config
  - name: declarative_config_string
{% endkong_config_table %}
<!--vale on-->

## DB-less mode with Kubernetes

You can run DB-less mode with Kubernetes both with and without {{ site.kic_product_name }}.

### DB-less mode with {{ site.kic_product_name }}

{{ site.kic_product_name }} provides a Kubernetes native way to configure {{ site.base_gateway }} using custom resource definitions (CRDs). In this deployment pattern, {{ site.base_gateway }} is deployed in DB-less mode, where the Data Plane configuration is held in memory.

> Operators configure {{ site.base_gateway }} using standard CRDs such as `Ingress` and `HTTPRoute`, and {{ site.kic_product_name }} translates those resources in to {{site.base_gateway}} entities before sending a request to update the running Data Plane configurations.

Reviewer: ```suggestion
Operators configure {{ site.base_gateway }} using standard CRDs such as `Ingress` and `HTTPRoute`, and {{ site.kic_product_name }} translates those resources into {{site.base_gateway}} entities before sending a request to update the running Data Plane configurations.
```

> In this topology, the Kubernetes API server is your source of truth. {{ site.kic_product_name }} reads resources stored on the API server and translates them in to a valid {{site.base_gateway}} configuration object. You can think of {{ site.kic_product_name }} as the Control Plane for your DB-less Data Planes.

Reviewer: ```suggestion
In this topology, the Kubernetes API server is your source of truth. {{ site.kic_product_name }} reads resources stored on the API server and translates them into a valid {{site.base_gateway}} configuration object. You can think of {{ site.kic_product_name }} as the Control Plane for your DB-less Data Planes.
```

Use `--set-file dblessConfig.config=/path/to/declarative-config.yaml` in Helm commands to substitute in a complete declarative config file.

Externally supplied ConfigMaps aren't hashed or tracked in deployment annotations. Subsequent ConfigMap updates require user-initiated deployment rollouts to apply the new configuration. Run `kubectl rollout restart deploy` after updating externally supplied ConfigMap content.

## DB-less mode limitations

There are a number of limitations you should be aware of when using {{site.base_gateway}} in DB-less
mode.

### Memory cache requirements

The entire configuration of entities must fit inside the {{site.base_gateway}}
cache. Make sure that the in-memory cache is configured appropriately in [`kong.conf`](/gateway/manage-kong-conf/):

<!--vale off-->
{% kong_config_table %}
config:
  - name: mem_cache_size
{% endkong_config_table %}
<!--vale on-->

### No central database coordination

> Since there is no central database, multiple {{site.base_gateway}} nodes have no

Reviewer: ```suggestion
Since there is no central database, {{site.base_gateway}} nodes have no
```

central coordination point and no cluster propagation of data.
Nodes are completely independent of each other.

This means that the declarative configuration should be loaded into each node
independently. Using the [`/config` endpoint](/api/gateway/admin-ee/#/operations/get-config) doesn't affect other {{site.base_gateway}}
nodes, since they have no knowledge of each other.

### Read-only Admin API

Since the only way to configure entities is via declarative configuration,
the endpoints for CRUD operations on entities are effectively read-only
in the [Admin API](/api/gateway/admin-ee/) when running {{site.base_gateway}} in DB-less mode. `GET` operations
for inspecting entities work as usual, but attempts to `POST`, `PATCH`
`PUT` or `DELETE` in endpoints such as `/services` or `/plugins` will return
`HTTP 405 Not Allowed`.

> This restriction is limited to what would be otherwise database operations. In

Reviewer: ```suggestion
This restriction is limited to what would otherwise be database operations. In
```

particular, using `POST` to set the health state of targets is still enabled,
since this is a node-specific in-memory operation.

### Kong Manager compatibility

> [Kong Manager](/gateway/kong-manager/) cannot guarantee compatibility with {{site.base_gateway}} operating in DB-less mode. You cannot create, update, or delete entities with Kong Manager when {{site.base_gateway}} is running in this mode. Entity counters in the **Summary** section on the global and workspace overview pages will not function correctly as well.

Reviewer: ```suggestion
[Kong Manager](/gateway/kong-manager/) cannot guarantee compatibility with {{site.base_gateway}} operating in DB-less mode. You cannot create, update, or delete entities with Kong Manager when {{site.base_gateway}} is running in this mode. Entity counters in the **Summary** section on the global and workspace overview pages will not function correctly either.
```
